[StabilizerTask] Compute net wrench only from force-sensors in contact

arntanguy opened the pull request:
Minor improvements to the `StabilizerTask` and `StabilizerStandingState`:

- Force sensor names were previously hard-coded. They are now determined automatically from the foot surfaces if a sensor is attached to them, or an exception is raised if no sensor is attached to the surface.
- `net wrench` measurements are now computed only from the force sensors in contact. This prevents undue noise from the sensors in the air during the swing foot phase.

- Added setters to StabilizerTask for all of the StabilizerConfig element. Previously one had to obtain a copy of the stabilizer configuration, modify it, and set the whole configuration again. This also fixes some minor issues when changing some of the tasks parameters. Notably for the contact tasks, changing the tasks parameters was delayed until the next contact change, it now occurs immediately as expected.

- Made the safety thresholds configurable (all the former MAX_*, MIN_* constexpr variables). Those values are somewhat arbitrary, and in some context it may make sense to modify them.

- Significantly improved the JSON schema. It now includes the default values from mc_rtc, with a note mentioning that they may be overridden by the RobotModule. It's quite useful to people to know roughly which tasks' weights to use in conjunction with the stabilizer. Previously one would have had to dig through the code, or dump the stabilizer configuration from the (slightly hidden) GUI.

- Added support for indirect force sensors. That's forward-thinking to using HRP4C toe-joint in the future, although that'll likely involve more changes (such as changing the contact surface used online).

- Added datastore setters to the StabilizerStandingState for the pelvis/torso/com tasks weights/stiffness.

- The default configuration of the states provided with the `StabilizerStandingState` (e.g `Stabilizer::Standing`, `Stabilizer::GoLeft`, etc) now target the contact surfaces frame defined in the stabilizer's configuration (foot center by default in the RobotModule) instead of the frame under the ankles. We have observed in practice on HRP robots that for the typical use-cases of this state (manipulation with external force perturbation), targetting the center of the foot instead of the ankle provides a higher stability-margin.
**Important: this may result in minor behavioural changes in existing controllers**.  The former behavior may be achieved with 
```yaml
Stabilizer::GoCenter:
  base: Stabilizer::Standing
  above: CenterAnkles
  completion:
    dcmEval: [0.005, 0.005, 0.05]

Stabilizer::GoLeft:
  base: Stabilizer::GoCenter
  above: LeftAnkle

Stabilizer::GoRight:
  base: Stabilizer::GoLeft
  above: RightAnkle
```

gergondet commented:
> Use LeftFoot/RightFoot surfaces instead of LeftFootCenter/RightFootCenter in the sample LIPMStabilizer FSM. This makes it compatible with more robots, including HRP-5P.

I recall a discussion with @caron that the distinction exists because one is located exactly below the force sensor (`LeftFoot`/`Rightfoot`) and the other is just centered in the sole polygon (`LeftFootCenter`/`RightFootCenter`). I am not sure whether this distinction is still important but if not, nevermind this comment and feel free to merge :)

If it's still important we can easily compute the centered frame from the other one so that we don't have to redefine for each new robot.

arntanguy commented:
I'll hold off a bit on this MR until I can test some additional things:
- I'd like to see the effect of the stabilizer when we abuse in in multi-contact with and without the hand sensors contributing to the net wrench. The way we currently use it for some multi-contact scenarios (wheel manipulation, etc) is already an abuse, let's see if this'd break or improve things. Ideally we shouldn't be using it in that way and have a better method implemented, that's coming, but in the meanwhile we need to keep supporting these edge cases ;)

- We noticed experimentally that it's better to give a reference to the centre of the foot for manipulation-like tasks. At least on HRP-2Kai and HRP4J, the ankles are pretty close to the back of the foot, which makes even slight perturbations in the sagittal direction hard/impossible to recover from. Targetting the centre of the foot is better in that case. As such, I'm thinking of adding default states to `StabilizerStanding` to either target the ankle or the foot's center, and making the center target the new default. The reasoning here is that the default YAML FSM behaviour will most likely be used in static cases, while more dynamic motions will most likely be implemented as specific C++ states.

arntanguy commented:
I've rebased this onto master and:

- Added setters to `StabilizerTask` for all of the `StabilizerConfig` element. Previously one had to obtain a copy of the stabilizer configuration, modify it, and set the whole configuration again. This also fixes some minor issues when changing some of the tasks parameters. Notably for the contact tasks, changing the tasks parameters was delayed until the next contact change, it now occurs immediately as expected.

- Made the safety thresholds configurable (all the former `MAX_*`, `MIN_*` constexpr variables). Those values are somewhat arbitrary, and in some context it may make sense to modify them.

- Significantly improved the JSON schema. It now includes the default values from `mc_rtc`, with a note mentioning that they may be overridden by the RobotModule. It's quite useful to people to know roughly which tasks' weights to use in conjunction with the stabilizer. Previously one would have had to dig through the code, or dump the stabilizer configuration from the (slightly hidden) GUI.

- Added support for indirect force sensors. That's forward-thinking to using HRP4C toe-joint in the future, although that'll likely involve more changes (such as changing the contact surface used online).

- Added datastore setters to the `StabilizerStandingState` for the pelvis/torso/com tasks weights/stiffness.

> I'd like to see the effect of the stabilizer when we abuse in in multi-contact with and without the hand sensors contributing to the net wrench.

Nevermind. This PR does not change the former behaviour, it merely replaces the hardcoded sensor names with non-hardcoded ones.

> We noticed experimentally that it's better to give a reference to the centre of the foot for manipulation-like tasks. 

In practice this is pretty much the only use case for the `StabilizerStandingState`, and we noticed that it was quite easy to make it fail by pushing itself over the back edge of the foot when manipulating with a bit too much external force. I think it's better to target the center of the foot by default for extra safety margin, as done by this PR. (This does not affect the behavior of LIPMWalking).


I have tested this and confirmed it with my modified version of LIPMWalking that uses this stabilizer and the observers from mc_rtc. Walking and stair climbing is working as expected.

@gergondet Could you have a quick look at these new changes?

gergondet commented on doc/_data/schemas/common/stabilizer_safety_thresholds.json:
```diff
@@ -0,0 +1,20 @@
+    "MIN_DS_PRESSURE": { "type": "number", "minimum": 0, "default": 15 },
+    "MIN_NET_TOTAL_FORCE_ZMP": { "type": "number", "minimum": 0, "default": 1 }
+  }
+}
```
Adding a reference to the documentation would be good to let people understand the meaning of these gains (not sure if `{% doxygen %}` will work in this context but it's worth the try)

arntanguy replied:
I don't think it does, I think I tried before, but I'll check.

arntanguy replied:
This took me a while, but it should work now, along with rendering any liquid tag from the schemas "description" elements.

gergondet commented on include/mc_tasks/lipm_stabilizer/StabilizerTask.h:
```diff
@@ -315,6 +301,167 @@ struct MC_TASKS_DLLAPI StabilizerTask : public MetaTask
+   * @brief Changes the safety thresholds
+   *
+   * This ensures that all the parameters depending on these safety parameters
+   * are withing the new thresholds. If they are out of bounds, they will be
```
*within

gergondet commented on include/mc_tasks/lipm_stabilizer/StabilizerTask.h:
```diff
@@ -315,6 +301,167 @@ struct MC_TASKS_DLLAPI StabilizerTask : public MetaTask
+  {
+    c_.safetyThresholds = thresholds;
+    c_.clampGains();
+    copAdmittance(c_.copAdmittance);
```
A comment to say that this is only required because we want to apply the new gain immediately would be nice

gergondet commented on src/mc_tasks/lipm_stabilizer/StabilizerTask.cpp:
```diff
@@ -329,6 +316,25 @@ void StabilizerTask::configure_(mc_solver::QPSolver & solver)
+  auto checkSurface = [&](const std::string & surfaceName) {
+    if(!robot().hasSurface(surfaceName))
+    {
+      mc_rtc::log::error_and_throw<std::runtime_error>("[{}] requires a surface named {} in robot {}", surfaceName,
```
This is missing one parameter

gergondet commented on src/mc_tasks/lipm_stabilizer/StabilizerTask.cpp:
```diff
@@ -329,6 +316,25 @@ void StabilizerTask::configure_(mc_solver::QPSolver & solver)
+    if(!robot().surfaceHasIndirectForceSensor(surfaceName))
+    {
+      mc_rtc::log::error_and_throw<std::runtime_error>(
+          "[StabilizerTask] Surface {} must have an associated force sensor.");
```
This is missing one parameter

gergondet commented:
> @gergondet Could you have a quick look at these new changes?

LGTM (modulo minor changes), I'll let you look into the CI failures :+1:
